### PR TITLE
Optimize lazy initializations 

### PR DESCRIPTION
**Pull request by kipply**
semantically, `||=` is "assign if", though it is usually used as "assign once".

These changes make it so that the interpreter is aware of this behaviour, and compiles the right hand side iff the assignment happens more than once. This saves some time and space! 

Fairly reliable [static profiling](https://gist.github.com/kipply/b64bdedc3c8accd869777642726adbc3) was done on 20 projects (2082 `||=` operators) and could safely confirm usage as lazy initialization 64% of the time.

### CODE REVIEW

**Comment by chrisseaton** [comment]
https://github.com/Shopify/truffleruby/issues/1

**Comment by chrisseaton** [comment]
Is this PR in CI? Is it waiting for anything?

**Comment by eregon** [comment]
For me :) Sorry, I forgot about it, I'll integrate it now.

File: src/main/java/org/truffleruby/language/control/OrLazyValueDefinedNode.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2013, 2020 Oracle and/or its affiliates. All rights reserved. This
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/language/control/OrLazyValueDefinedNode.java]: Why does this copyright go back to 2013? Is the code copied from another file?

**Comment by eregon** [comment]
Merged in 8db21a3d1d01fccdc50751a00d20e745a9e9a527, thank you for the PR!

We're thinking to generalize this pattern and have some kind of ConditionProfile which just doesn't profile the first execution of a branch or so. It could likely be useful in other places where there is a lazy computation that is likely to happen only once.

**Pull request merged**